### PR TITLE
HIVE-27499: Include optional db/table names in the NotificationEvent …

### PR DESCRIPTION
**Pull request by saihemanth-cloudera**
…thrift object

### What changes were proposed in this pull request?
Made some modifications in the thrift object to include optional db, table names in NotificationEventRequest object(s).

### Why are the changes needed?
External clients can fetch the pending events based on given db, table names.

### Does this PR introduce _any_ user-facing change?
No.

### Is the change a dependency upgrade?
No.

### How was this patch tested?
Local environment. Added unit test to validate the feature.

### CODE REVIEW

**Comment by saihemanth-cloudera** [comment]
@henrib @VenuReddy2103 -- Can you review this PR? Thanks.

File: standalone-metastore/metastore-server/src/main/sql/mssql/upgrade-4.0.0-alpha-2-to-4.0.0-beta-1.mssql.sql [hunk 1]
```diff
@@ -44,6 +44,8 @@ UPDATE "SERDES"
             LEFT JOIN "SDS" ON "TBLS"."SD_ID" = "SDS"."SD_ID"
             WHERE "TBL_ID" IN (SELECT "TBL_ID" FROM "TABLE_PARAMS" WHERE "PARAM_VALUE" LIKE '%KuduStorageHandler%')
     );
+-- HIVE-27499
+CREATE UNIQUE INDEX NOTIFICATION_LOG_UNIQUE_DB ON NOTIFICATION_LOG (DB_NAME, EVENT_ID);
```
Comment (dengzhhu653) [inline standalone-metastore/metastore-server/src/main/sql/mssql/upgrade-4.0.0-alpha-2-to-4.0.0-beta-1.mssql.sql:48]: Is it possible to add table name, catalog name to the index?

Comment (henrib) [reply standalone-metastore/metastore-server/src/main/sql/mssql/upgrade-4.0.0-alpha-2-to-4.0.0-beta-1.mssql.sql:48]: I believe this new index is helping selectivity for notification log queries qualified by a database name and a range on event_id (if not mistaken). Adding table,catalog name would not help this purpose.

Comment (dengzhhu653) [reply standalone-metastore/metastore-server/src/main/sql/mssql/upgrade-4.0.0-alpha-2-to-4.0.0-beta-1.mssql.sql:48]: I couldn't find searching the notification by the range of event_id in the changes, am I miss something?

Comment (saihemanth-cloudera) [reply standalone-metastore/metastore-server/src/main/sql/mssql/upgrade-4.0.0-alpha-2-to-4.0.0-beta-1.mssql.sql:48]: > Is it possible to add table name, catalog name to the index? looks like the EVENT_ID is helpless when using the index
The idea of using an index with DB_NAME, EVENT_ID is that, requirement from the HMS/Impala client is that they would like to know what are latest events in the notification_log table from the specified event id in a particular database or for a specified table(s) in the database. 
So following are the fields that would be specified fromEventId, database name, catalog name, table name(s).
Coming to your question, does it help adding table name and catalog name to the index, I think it would worsen the performance of index because notification_log table would have an entry every second on average basis and every single entry the index would have to be updated. So it becomes costly to maintain a composite key will the above fields. Instead, if we have an index only on database name and event id, it is less costly to manage the index and much more efficient to required info.

> I couldn't find searching the notification by the range of event_id in the changes, am I miss something?
range of event_ids is already supported. no changes are required in that perspective. Above mentioned is the required use case.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/ObjectStore.java
Comment (henrib) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/ObjectStore.java:1]: The logic that generates parameter names and add their values to the parameterVals might be easier to grasp with a comment and/or adding a variable to hold the parameter 'number' (parameterVals.size()).

**Review by henrib — CHANGES_REQUESTED** [review]
Query generation could use a comment and/or local variables.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/ObjectStore.java
Comment (saihemanth-cloudera) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/ObjectStore.java:1]: parameterVals.size() may keep changing after every conditional statement. We'll have to assign the size to a variable every time it changes. So I think it is OK to leave it as it is.

Comment (saihemanth-cloudera) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/ObjectStore.java:1]: Ack

**Review by nrg4878 — CHANGES_REQUESTED** [review]

File: standalone-metastore/metastore-server/src/main/sql/derby/hive-schema-4.0.0-beta-1.derby.sql [hunk 2]
```diff
@@ -189,6 +189,8 @@ CREATE TABLE "APP"."NOTIFICATION_LOG" (
 
 CREATE UNIQUE INDEX "APP"."NOTIFICATION_LOG_EVENT_ID" ON "APP"."NOTIFICATION_LOG" ("EVENT_ID");
 
+CREATE UNIQUE INDEX "APP"."NOTIFICATION_LOG_UNIQUE_DB" ON "APP"."NOTIFICATION_LOG" ("DB_NAME", "EVENT_ID");
```
[the remark this answers is not in the document]
Comment (saihemanth-cloudera) [reply standalone-metastore/metastore-server/src/main/sql/derby/hive-schema-4.0.0-beta-1.derby.sql:192]: Both indices are required. There can be two different types of frequent queries in this table.
1) provided eventId, fetch all the events greater than the eventId.
2) provided eventId, dbName, table, fetch all the events of this table greater than the eventId.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/ObjectStore.java [hunk 3]
```diff
@@ -11831,8 +11831,35 @@ public NotificationEventResponse getNextNotification(NotificationEventRequest rq
...
+        filterBuilder.append(" && dbName == para" + parameterVals.size());
+      }
+      if (rqst.isSetTableNames()) {
+        filterBuilder.append(" && ");
+        for (String tableName : rqst.getTableNames()) {
+          parameterVals.add(tableName);
+          parameterBuilder.append(", java.lang.String para" + parameterVals.size());
+          filterBuilder.append("tableName == para" + parameterVals.size()+ " || ");
+        }
+        filterBuilder.setLength(filterBuilder.length() - 4); // remove the last " || "
```
Comment (dengzhhu653) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/ObjectStore.java:11861]: I think we should add parentheses to the `tableName` filter

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/ObjectStore.java [hunk 4]
```diff
@@ -11831,8 +11831,35 @@ public NotificationEventResponse getNextNotification(NotificationEventRequest rq
...
+        parameterVals.add(rqst.getCatName());
+        parameterBuilder.append(", java.lang.String para" + parameterVals.size());
+        filterBuilder.append(" && catalogName == para" + parameterVals.size());
+      }
+      if (rqst.isSetDbName()) {
+        parameterVals.add(rqst.getDbName());
+        parameterBuilder.append(", java.lang.String para" + parameterVals.size());
+        filterBuilder.append(" && dbName == para" + parameterVals.size());
+      }
+      if (rqst.isSetTableNames()) {
```
Comment (dengzhhu653) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/ObjectStore.java:11854]: should we check to see if `rqst.getTableNames()` is empty?

File: itests/hcatalog-unit/src/test/java/org/apache/hive/hcatalog/listener/TestDbNotificationListener.java [hunk 5]
```diff
@@ -1684,6 +1684,58 @@ public void cleanupNotifs() throws Exception {
...
+    assertEquals(24, msClient.getNotificationEventsCount(eventsReq).getEventsCount());
+    eventsReq.unsetTableNames();
+    assertEquals(37, msClient.getNotificationEventsCount(eventsReq).getEventsCount());
+  }
+
+  private void generateSometableEvents(String dbName, String tableName) throws Exception {
+    // Event 1
+    driver.run("create database if not exists "+dbName);
+    // Event 2
+    driver.run("use "+dbName);
```
Comment (dengzhhu653) [inline itests/hcatalog-unit/src/test/java/org/apache/hive/hcatalog/listener/TestDbNotificationListener.java:1726]: we do not generate an event for `use db`?

File: standalone-metastore/metastore-server/src/main/sql/mssql/hive-schema-4.0.0-beta-2.mssql.sql [hunk 6]
```diff
@@ -661,6 +661,8 @@ ALTER TABLE NOTIFICATION_LOG ADD CONSTRAINT NOTIFICATION_LOG_PK PRIMARY KEY (NL_
 
 CREATE UNIQUE INDEX NOTIFICATION_LOG_EVENT_ID ON NOTIFICATION_LOG (EVENT_ID);
 
+CREATE UNIQUE INDEX NOTIFICATION_LOG_UNIQUE_DB ON NOTIFICATION_LOG (DB_NAME, EVENT_ID);
```
Comment (dengzhhu653) [inline standalone-metastore/metastore-server/src/main/sql/mssql/hive-schema-4.0.0-beta-2.mssql.sql:664]: As we have already indexed on `EVENT_ID` for this table, do we need to create a composite index on (DB_NAME, EVENT_ID)?

Comment (saihemanth-cloudera) [reply standalone-metastore/metastore-server/src/main/sql/mssql/hive-schema-4.0.0-beta-2.mssql.sql:664]: Both indices are required. There can be two different types of frequent queries in this table.

1. provided eventId, fetch all the events greater than the eventId. -- Will use index with event_id 
2. provided eventId, dbName, table, fetch all the events of this table greater than the eventId. -- Will use index with db_name and event_id

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/ObjectStore.java [hunk 7]
```diff
@@ -11856,8 +11856,35 @@ public NotificationEventResponse getNextNotification(NotificationEventRequest rq
...
+        parameterVals.add(rqst.getCatName());
+        parameterBuilder.append(", java.lang.String para" + parameterVals.size());
+        filterBuilder.append(" && catalogName == para" + parameterVals.size());
+      }
+      if (rqst.isSetDbName()) {
+        parameterVals.add(rqst.getDbName());
+        parameterBuilder.append(", java.lang.String para" + parameterVals.size());
+        filterBuilder.append(" && dbName == para" + parameterVals.size());
+      }
+      if (rqst.isSetTableNames() && !rqst.getTableNames().isEmpty()) {
```
Comment (dengzhhu653) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/ObjectStore.java:11879]: Sorry, I mean we should do like this:
 `dbName == para2 && tableName == para3  || tableName == para4 && eventType != para5` 
to
  `dbName == para2 && (tableName == para3  || tableName == para4) && eventType != para5`

File: standalone-metastore/metastore-common/src/gen/thrift/gen-javabean/org/apache/hadoop/hive/metastore/api/AlterPartitionsRequest.java [hunk 8]
```diff
@@ -1028,14 +1028,14 @@ public void read(org.apache.thrift.protocol.TProtocol iprot, AlterPartitionsRequ
           case 4: // PARTITIONS
             if (schemeField.type == org.apache.thrift.protocol.TType.LIST) {
               {
-                org.apache.thrift.protocol.TList _list1360 = iprot.readListBegin();
-                struct.partitions = new java.util.ArrayList<Partition>(_list1360.size);
-                @org.apache.thrift.annotation.Nullable Partition _elem1361;
-                for (int _i1362 = 0; _i1362 < _list1360.size; ++_i1362)
+                org.apache.thrift.protocol.TList _list1376 = iprot.readListBegin();
```
Comment (dengzhhu653) [inline standalone-metastore/metastore-common/src/gen/thrift/gen-javabean/org/apache/hadoop/hive/metastore/api/AlterPartitionsRequest.java:1031]: we don't make any changes on the AlterPartitionsRequest, not sure how these comes from?

File: standalone-metastore/metastore-server/src/main/sql/mssql/hive-schema-4.0.0-beta-2.mssql.sql (hunk 6 above)
Comment (dengzhhu653) [reply standalone-metastore/metastore-server/src/main/sql/mssql/hive-schema-4.0.0-beta-2.mssql.sql:664]: I'm wondering for some back db, for the second case: event_id > 10 and db_name = 'default'
 1. use index with event_id to filter the matched rows (event_id > 10)
 2. on the result of the first step, apply the index with db_name (db_name = 'default')
if the event_id > 10 is only return one row from 1000 rows, then this is better than using the index with db_name and event_id.
As the API is aim to fetch the latest event in most cases, so I think this is common in real production.

Comment (saihemanth-cloudera) [reply standalone-metastore/metastore-server/src/main/sql/mssql/hive-schema-4.0.0-beta-2.mssql.sql:664]: Any DBMS would use one and only one index to query the result. So depending on the query on the notification log table, it has to use index 1 or index 2.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/ObjectStore.java (hunk 7 above)
Comment (saihemanth-cloudera) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/ObjectStore.java:11879]: Ack

File: standalone-metastore/metastore-common/src/gen/thrift/gen-javabean/org/apache/hadoop/hive/metastore/api/AlterPartitionsRequest.java (hunk 8 above)
Comment (saihemanth-cloudera) [reply standalone-metastore/metastore-common/src/gen/thrift/gen-javabean/org/apache/hadoop/hive/metastore/api/AlterPartitionsRequest.java:1031]: Thrift file generation is little complex to understand. I believe it has an order for the fields in the thrift objects. if we insert a new field in one of the thrift objects, it re-generates all other thrift objects also.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/ObjectStore.java [hunk 9]
```diff
@@ -11856,8 +11856,36 @@ public NotificationEventResponse getNextNotification(NotificationEventRequest rq
...
+      ->  parameterBuilder: java.lang.Long para0, java.lang.String para1, java.lang.String para2
+          , java.lang.String para3, java.lang.String para4, java.lang.String para5
+       */
+      if (rqst.isSetCatName()) {
+        parameterVals.add(rqst.getCatName());
+        parameterBuilder.append(", java.lang.String para" + parameterVals.size());
+        filterBuilder.append(" && catalogName == para" + parameterVals.size());
+      }
+      if (rqst.isSetDbName()) {
+        parameterVals.add(rqst.getDbName());
```
Comment (dengzhhu653) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/ObjectStore.java:11875]: nit: nit: normalizeIdentifier(rqst.getDbName())

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/ObjectStore.java [hunk 10]
```diff
@@ -11856,8 +11856,36 @@ public NotificationEventResponse getNextNotification(NotificationEventRequest rq
...
+        parameterBuilder.append(", java.lang.String para" + parameterVals.size());
+        filterBuilder.append(" && dbName == para" + parameterVals.size());
+      }
+      if (rqst.isSetTableNames() && !rqst.getTableNames().isEmpty()) {
+        filterBuilder.append(" && (");
+        for (String tableName : rqst.getTableNames()) {
+          parameterVals.add(tableName);
```
Comment (dengzhhu653) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/ObjectStore.java:11882]: nit: nit: normalizeIdentifier(tableName)

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/ObjectStore.java [hunk 11]
```diff
@@ -11856,8 +11856,36 @@ public NotificationEventResponse getNextNotification(NotificationEventRequest rq
       long lastEvent = rqst.getLastEvent();
       List<Object> parameterVals = new ArrayList<>();
       parameterVals.add(lastEvent);
+      // filterBuilder parameter is used for construction of conditional clause in the select query
       StringBuilder filterBuilder = new StringBuilder("eventId > para" + parameterVals.size());
+      // parameterBuilder parameter is used for specify what types of parameters will go into the filterBuilder
       StringBuilder parameterBuilder = new StringBuilder("java.lang.Long para" + parameterVals.size());
+      /* A fully constructed query would like:
```
Comment (dengzhhu653) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/ObjectStore.java:11863]: nit: need to update the comments

**Comment by dengzhhu653** [comment]
Leave some minor comments, others look fine to me. I need to take a more look on the index, cc @nrg4878 @henrib any thoughts about the added index?

File: standalone-metastore/metastore-server/src/main/sql/mssql/hive-schema-4.0.0-beta-2.mssql.sql (hunk 6 above)
Comment (VenuReddy2103) [reply standalone-metastore/metastore-server/src/main/sql/mssql/hive-schema-4.0.0-beta-2.mssql.sql:664]: I agree with @dengzhhu653 
Optimizer would check the potential indexes available and decides to choose one(could be single index or composite index) among them or not to choose any index based on various factors. Since we have index with single column(`EVENT_ID`), in query having where clause with eventId + may be dbName + may be table name etc, optimizer detects the eventId index as potential candidate. Have captured explain queries on mysql for index used and not used cases.
Index used:
```
mysql> explain select * from NOTIFICATION_LOG where db_name='mydb' and event_id>8;
+----+-------------+------------------+------------+-------+---------------------------+---------------------------+---------+------+------+----------+------------------------------------+
| id | select_type | table            | partitions | type  | possible_keys             | key                       | key_len | ref  | rows | filtered | Extra                              |
+----+-------------+------------------+------------+-------+---------------------------+---------------------------+---------+------+------+----------+------------------------------------+
|  1 | SIMPLE      | NOTIFICATION_LOG | NULL       | range | NOTIFICATION_LOG_EVENT_ID | NOTIFICATION_LOG_EVENT_ID | 8       | NULL |    3 |    10.00 | Using index condition; Using where |
+----+-------------+------------------+------------+-------+---------------------------+---------------------------+---------+------+------+----------+------------------------------------+
1 row in set, 1 warning (0.00 sec)

mysql> explain select * from NOTIFICATION_LOG where db_name='mydb' and tbl_name='stage' and event_id>8;
+----+-------------+------------------+------------+-------+---------------------------+---------------------------+---------+------+------+----------+------------------------------------+
| id | select_type | table            | partitions | type  | possible_keys             | key                       | key_len | ref  | rows | filtered | Extra                              |
+----+-------------+------------------+------------+-------+---------------------------+---------------------------+---------+------+------+----------+------------------------------------+
|  1 | SIMPLE      | NOTIFICATION_LOG | NULL       | range | NOTIFICATION_LOG_EVENT_ID | NOTIFICATION_LOG_EVENT_ID | 8       | NULL |    3 |     9.09 | Using index condition; Using where |
+----+-------------+------------------+------------+-------+---------------------------+---------------------------+---------+------+------+----------+------------------------------------+
1 row in set, 1 warning (0.00 sec)

```
Index doesn't seem to be used here:
```
mysql> explain select * from NOTIFICATION_LOG where event_id>12 and db_name='mydb';
+----+-------------+------------------+------------+------+---------------------------+------+---------+------+------+----------+-------------+
| id | select_type | table            | partitions | type | possible_keys             | key  | key_len | ref  | rows | filtered | Extra       |
+----+-------------+------------------+------------+------+---------------------------+------+---------+------+------+----------+-------------+
|  1 | SIMPLE      | NOTIFICATION_LOG | NULL       | ALL  | NOTIFICATION_LOG_EVENT_ID | NULL | NULL    | NULL |   29 |     5.86 | Using where |
+----+-------------+------------------+------------+------+---------------------------+------+---------+------+------+----------+-------------+
1 row in set, 1 warning (0.00 sec)
```
Have found some reference from mysql but this seem to be not very specific to mysql.
https://dev.mysql.com/doc/refman/8.0/en/multiple-column-indexes.html#:~:text=If%20the%20table%20has%20a%20multiple,of%20(col1%2C%20col2%2C%20col3).
https://dev.mysql.com/doc/refman/8.0/en/index-condition-pushdown-optimization.html#:~:text=Get%20the%20next%20row%27s,on%20the%20test%20result.

Some dbs seem to provide a way to specify index hints in query explicilty to always use an index. 

If we add the dbname + eventId composite index, index updation/maintence overhead may be high because index data need to be sorted by dbname first and then by eventIds within the db on the huge notification log? It may improve the query performance though.

May be good to compare the tradeoffs between using the eventId based index vs dbname + eventId composite index ?

Comment (saihemanth-cloudera) [reply standalone-metastore/metastore-server/src/main/sql/mssql/hive-schema-4.0.0-beta-2.mssql.sql:664]: @VenuReddy2103 - Good finding. The index I introduced is only used if there is event_id value is absolute
`mysql> explain analyze select * from NOTIFICATION_LOG where db_name='mydb';
+-----------------------------------------------------------------------------------------------------------------------------------------------------+
| EXPLAIN                                                                                                                                             |
+-----------------------------------------------------------------------------------------------------------------------------------------------------+
| -> Index lookup on NOTIFICATION_LOG using NOTIFICATION_LOG_UNIQUE_DB (DB_NAME='mydb')  (cost=0.9 rows=4) (actual time=0.276..0.738 rows=4 loops=1)
 |
+-----------------------------------------------------------------------------------------------------------------------------------------------------+`
But the real use case query would definitely have a range of conditions on the event_id (eg: event_id>100), but apparently the first index is being used 'NOTIFICATION_LOG_EVENT_ID' instead of 'NOTIFICATION_LOG_UNIQUE_DB'. This is something not desired.
`mysql> explain analyze select * from NOTIFICATION_LOG where db_name='mydb' and (tbl_name='tb1' or tbl_name='tb2') and event_id>1;
+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| EXPLAIN                                                                                                                                                                                                                                                                                                                                                                                                                      |
+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| -> Filter: ((notification_log.DB_NAME = 'mydb') and ((notification_log.TBL_NAME = 'tb1') or (notification_log.TBL_NAME = 'tb2')))  (cost=1.61 rows=1.31) (actual time=0.0784..0.0969 rows=2 loops=1)
    -> Index range scan on NOTIFICATION_LOG using NOTIFICATION_LOG_EVENT_ID over (1 < EVENT_ID), with index condition: (notification_log.EVENT_ID > 1)  (cost=1.61 rows=3) (actual time=0.0731..0.0902 rows=3 loops=1)
 |
+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+`
So in summary the second index I intend to introduce has no effect. There is no point in forcing the DBMS to use the index because it is inefficient in reading the results and has an overhead to maintain the second index. I would rather remove the newly introduced index. Thanks

File: itests/hcatalog-unit/src/test/java/org/apache/hive/hcatalog/listener/TestDbNotificationListener.java [hunk 12]
```diff
@@ -117,8 +117,17 @@
  * Tests DbNotificationListener when used as a transactional event listener
  * (hive.metastore.transactional.event.listeners)
  */
+<<<<<<< HEAD
```
Comment (dengzhhu653) [inline itests/hcatalog-unit/src/test/java/org/apache/hive/hcatalog/listener/TestDbNotificationListener.java:120]: can we remove these conflict codes?

File: standalone-metastore/metastore-server/src/main/sql/postgres/upgrade-4.0.0-beta-1-to-4.0.0-beta-2.postgres.sql [hunk 13]
```diff
@@ -1,4 +1,5 @@
 SELECT 'Upgrading MetaStore schema from 4.0.0-beta-1 to 4.0.0-beta-2';
+
```
Comment (dengzhhu653) [inline standalone-metastore/metastore-server/src/main/sql/postgres/upgrade-4.0.0-beta-1-to-4.0.0-beta-2.postgres.sql:2]: nit: extra line

**Review by dengzhhu653 — APPROVED** [review]

**Review by nrg4878 — APPROVED** [review]
Thank you for the changes to the test. Looks good to me. +1
